File: backend/src/ingestion/einvoice_loader.ts
import db from '../database';

export async function loadEInvoice(data: any): Promise<any> {
    const session = db.getSession();
    try {
        const tx = session.beginTransaction();

        // Create IRN node
        await tx.run(`
      MERGE (irn:IRN {irnHash: $irnHash})
      ON CREATE SET irn.status = $status, irn.generationDate = $generationDate
      ON MATCH SET irn.status = $status
    `, {
            irnHash: data.irn,
            status: data.status,
            generationDate: data.generationDate
        });

        // Link Invoice to IRN
        await tx.run(`
      MERGE (i:Invoice {invoiceNo: $invoiceNo, supplierGstin: $supplierGstin})
      MERGE (i)-[:HAS_IRN]->(irn:IRN {irnHash: $irnHash})
    `, {
            invoiceNo: data.invoiceDetails.invoiceNo,
            supplierGstin: data.supplierGstin,
            irnHash: data.irn
        });

        await tx.commit();
        return { status: "success", irn: data.irn };
    } catch (error) {
        console.error("e-Invoice Load Error:", error);
        throw error;
    } finally {
        await session.close();
    }
}

File: backend/app/ingestion/einvoice_loader.py
"""E-Invoice loader — IRN node ingestion into the knowledge graph."""

from app.database import get_session


def ingest_einvoice(data: dict) -> dict:
    """
    Parse e-invoice JSON payload and create/update:
    - IRN nodes
    - HAS_IRN edges from Invoice -> IRN
    """
    irn_entries = data["irn_entries"]
    stats = {"irn_merged": 0}
    
    with get_session() as session:
        for entry in irn_entries:
            # MERGE IRN node
            session.run(
                """
                MERGE (irn:IRN {irnHash: $irnHash})
                ON CREATE SET irn.status = $status, irn.invoiceNo = $invoiceNo,
                              irn.generatedDate = $generatedDate
                ON MATCH SET irn.status = $status
                """,
                irnHash=entry["irnHash"],
                status=entry["status"],
                invoiceNo=entry["invoiceNo"],
                generatedDate=entry.get("generatedDate"),
            )
            
            # HAS_IRN relationship: Invoice -> IRN
            session.run(
                """
                MATCH (i:Invoice {invoiceNo: $invoiceNo})
                MATCH (irn:IRN {irnHash: $irnHash})
                MERGE (i)-[:HAS_IRN]->(irn)
                """,
                invoiceNo=entry["invoiceNo"],
                irnHash=entry["irnHash"],
            )
            stats["irn_merged"] += 1
    
    return {
        "status": "success",
        "message": f"E-invoice data ingested: {stats['irn_merged']} IRNs",
        "stats": stats,
    }
